index.d.cts

```diff
@@ -31,10 +31,12 @@ type PrettierSynchronizedFunction<
   ...args: Parameters<Prettier[FunctionName]>
 ) => Awaited<ReturnType<Prettier[FunctionName]>>;
 
+export const createSynchronizedPrettier: (options: {
+  prettierEntry: string | URL;
+}) => SynchronizedPrettier;
+
 declare const synchronizedPrettier: SynchronizedPrettier & {
-  createSynchronizedPrettier: (options: {
-    prettierEntry: string | URL;
-  }) => SynchronizedPrettier;
+  createSynchronizedPrettier: typeof createSynchronizedPrettier;
 };
 
-export = synchronizedPrettier;
+export default synchronizedPrettier;
```

Comment:
Yeah you're right 😣

The PR works great with `{ "type": "module" }` but do you know any other ways to error on named exports?

```mjs
import { format } from '@prettier/sync'
```

I don't want to advertise a non-existent named synthetic `default`

Reply:
Unfortunately no. :(

The solution in https://github.com/prettier/prettier-synchronized/pull/12#pullrequestreview-1533120452 works, but it’s a hack.

Reply:
Actually, I forgot the official answer

Remember I mentioned TypeScript [`{ "resolvePackageJsonExports": true }`](https://www.typescriptlang.org/tsconfig#resolvePackageJsonExports) in https://github.com/prettier/prettier-synchronized/pull/10#discussion_r1263662985?

The non-hack approach would be to serve different types for `import` versus `require()` using:

```json
"exports": {
  ".": {
    "import": {
      "types": "./index.d.ts",
      "default": "./index.cjs"
    },
    "require": {
      "types": "./index.d.cts",
      "default": "./index.cjs"
    }
  },
  "./*": "./*"
},
```

Works great but you might be able to advise on compatibility?

Reply:
This would be sufficient:

```json
{
  "exports": {
    "types": {
      "import": "./index.d.ts",
      "default": "./index.d.cts"
    },
    "default": "./index.cjs"
  }
}
```

The thing is, this is also a hack. Each JavaScript file should be accompanied by one matching type definitions file. This adds type definitions for a JavaScript file that doesn’t exist, then makes it work using package exports.

I like the creative thinking though!

Another solution would be to not look at the type definitions, but rewrite the source code so that each member is rewritten as `exports.{key} = {value}`, i.e. `exports.format = format`. This means it would actually be possible to use named imports.

---

Are you running into any practical issues with the current solution though? For me this just works, and IMO it’s good enough as-is.

Reply:
Yeah, the current type definitions make it look like named exports are supported 😣

But try to use them in ESM code (not transpiled) and you'll see Node.js error with:

```console
import { format } from '@prettier/sync'
         ^^^^^^
SyntaxError: Named export 'format' not found. The requested module '@prettier/sync' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@prettier/sync';
const { format } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)

Node.js v18.16.1
```

Some Prettier properties are getters so ["default only" is by design](https://github.com/alphagov/govuk-frontend/pull/3971#discussion_r1265398678)